nuxeo-runtime/nuxeo-runtime-test/src/main/java/org/nuxeo/runtime/test/RuntimeHarnessImpl.java

```diff
@@ -0,0 +1,478 @@
+/*
+ * (C) Copyright 2006-2018 Nuxeo (http://nuxeo.com/) and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Nuxeo - initial API and implementation
+ */
+package org.nuxeo.runtime.test;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.Set;
+import java.util.jar.Attributes;
+import java.util.jar.Manifest;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.nuxeo.common.Environment;
+import org.nuxeo.osgi.BundleFile;
+import org.nuxeo.osgi.BundleImpl;
+import org.nuxeo.osgi.DirectoryBundleFile;
+import org.nuxeo.osgi.JarBundleFile;
+import org.nuxeo.osgi.OSGiAdapter;
+import org.nuxeo.osgi.SystemBundle;
+import org.nuxeo.osgi.SystemBundleFile;
+import org.nuxeo.osgi.application.StandaloneBundleLoader;
+import org.nuxeo.runtime.RuntimeServiceException;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.model.Extension;
+import org.nuxeo.runtime.model.RegistrationInfo;
+import org.nuxeo.runtime.model.RuntimeContext;
+import org.nuxeo.runtime.model.StreamRef;
+import org.nuxeo.runtime.model.URLStreamRef;
+import org.nuxeo.runtime.model.impl.DefaultRuntimeContext;
+import org.nuxeo.runtime.osgi.OSGiRuntimeContext;
+import org.nuxeo.runtime.osgi.OSGiRuntimeService;
+import org.nuxeo.runtime.test.protocols.inline.InlineURLFactory;
+import org.nuxeo.runtime.test.runner.RuntimeHarness;
+import org.nuxeo.runtime.test.runner.TargetExtensions;
+import org.nuxeo.runtime.transaction.TransactionHelper;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.FrameworkEvent;
+
+import io.github.lukehutch.fastclasspathscanner.FastClasspathScanner;
+
+public final class RuntimeHarnessImpl implements RuntimeHarness {
```

Comment:
Could you add javadoc with `@since` and remove final keyword ?

Reply:
Done